src/support/debug.cpp

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2015 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "support/debug.h"
+#include <execinfo.h>
+
+#include <err.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+// Adapted from:
+//  http://spin.atomicobject.com/2013/01/13/exceptions-stack-traces-c/
+
+namespace wasm {
+const char *globalProgramName = nullptr;
+
+#ifdef _WIN32
+void setDebugSignalHandler(const char *programName) {}
+#else
+// Resolve symbol name and source location given the path to the executable
+// and an address. For this to work you'll need to compile with:
+//  on linux:   gcc -g
+//  on OS X:    gcc -g -fno-pie
+int addr2line(char const *const program_name, void const *const addr) {
+  char cmd[512] = {0};
+#ifdef __APPLE__
+  sprintf(cmd, "atos -o %.256s %p", program_name, addr);
+#else
+  sprintf(cmd, "addr2line -f -p -e %.256s %p", program_name, addr);
+#endif
+  return system(cmd);
+}
+
+#define MAX_STACK_FRAMES 64
+static void *stackTraces[MAX_STACK_FRAMES];
+
+void printStackTrace(int offset = 5) {
+  if (!globalProgramName) {
+    printf("You must first call setDebugSignalHandler().\n");
+    return;
+  }
+  int i, traceSize = 0;
+  char **messages = (char **)nullptr;
+
+  traceSize = backtrace(stackTraces, MAX_STACK_FRAMES);
+  messages = backtrace_symbols(stackTraces, traceSize);
+  
+  // TODO: Call addr2line once with all the addresses, as it's much faster.
+  // Skip the first couple stack frames (as they are this function and
+  // our handler) and also skip the last frame as it's (always?) junk.
+  for (i = offset; i < (traceSize - 1); ++i) {
+    if (addr2line(globalProgramName, stackTraces[i]) != 0) {
+      printf("  error determining line # for: %s\n", messages[i]);
+    }
+  }
+  if (messages) {
+    free(messages);
+  }
+}
+
+void signalHandler(int sig, siginfo_t *siginfo, void *context) {
+  switch (sig) {
```

Comment:
I'm usually a fan of using xmacros for this type of repetitive table-like stuff :-)
Have an external `.inc` file with macro-ized things like `DO(SIGNUM, MESSAGE)`, and then `#define DO(SIGNUM, MESSAGE, ...) ...` and `#include "signals.inc"`.